lib/std/Build.zig

```diff
@@ -1285,10 +1285,14 @@ pub fn standardTargetOptions(b: *Build, args: StandardTargetOptionsArgs) Resolve
 }
 
 pub fn parseTargetQuery(options: std.Target.Query.ParseOptions) error{ParseFailed}!std.Target.Query {
-    var diags: Target.Query.ParseOptions.Diagnostics = .{};
-    var opts_copy = options;
-    opts_copy.diagnostics = &diags;
-    return std.Target.Query.parse(options) catch |err| switch (err) {
+    var diags_dummy: std.Target.Query.ParseOptions.Diagnostics = undefined;
+    const diags = options.diagnostics orelse &diags_dummy;
+    var opts_with_diags = options;
+    if (opts_with_diags.diagnostics == null) {
+        diags_dummy = .{};
+        opts_with_diags.diagnostics = diags;
+    }
+    return std.Target.Query.parse(opts_with_diags) catch |err| switch (err) {
```

Comment:
```suggestion
    // If the user passes in diagnostics, it indicates they want to take
    // responsibility for handling the ParseFailed error.
    if (options.diagnostics) |diags| return std.Target.Query.parse(options);

    var diags: std.Target.Query.ParseOptions.Diagnostics = .{};
    var opts_with_diags = options;
    opts_with_diags.diagnostics = &diags;
    return std.Target.Query.parse(opts_with_diags) catch |err| switch (err) {
```

Reply:
Hmm this suggestion suffers from an error set mismatch, which indicates some ambiguity regarding the user's intent.

I have a different counter suggestion then:

```diff
--- a/lib/std/Build.zig
+++ b/lib/std/Build.zig
@@ -1284,11 +1284,16 @@ pub fn standardTargetOptions(b: *Build, args: StandardTargetOptionsArgs) Resolve
     return b.resolveTargetQuery(query);
 }
 
+/// Obtain a target query from a string, reporting diagnostics to stderr if the
+/// parsing failed.
+/// Asserts that the `diagnostics` field of `options` is `null`. This use case
+/// is handled instead by calling `std.Target.Query.parse` directly.
 pub fn parseTargetQuery(options: std.Target.Query.ParseOptions) error{ParseFailed}!std.Target.Query {
+    assert(options.diagnostics == null);
     var diags: Target.Query.ParseOptions.Diagnostics = .{};
     var opts_copy = options;
     opts_copy.diagnostics = &diags;
-    return std.Target.Query.parse(options) catch |err| switch (err) {
+    return std.Target.Query.parse(opts_copy) catch |err| switch (err) {
         error.UnknownCpuModel => {
             std.debug.print("unknown CPU: '{s}'\navailable CPUs for architecture '{s}':\n", .{
                 diags.cpu_name.?, @tagName(diags.arch.?),
```

Reply:
I agree, good perspective for me, I am rarely critical of user intent in this way! Seems obvious now.